src/webgpu/shader/execution/builtin/builtin.ts

```diff
@@ -845,3 +847,80 @@ export const kValue = {
     toMinus32: -Math.pow(2, -32),
   },
 } as const;
+
+/**
+ * Checks for equality between two variable, doing a deep comparison if needed.
+ *
+ * |funcs_equal| controls how functions are compared. If |funcs_equal| is true,
+ * then functions are not deep inspected, and lhs and rhs both being a function
+ * is sufficient to pass.
+ * If false, then JSON.stringify is called to confirm they have the same body.
+ * */
+export function deepEquals(lhs: any, rhs: any, funcs_equal: boolean = false): boolean {
```

Comment:
I'm not familiar enough with TS or the testing framework to properly review this. My gut response is that I'm surprised that this is needed, and I'm not confident it'll work for all inputs. @kainino0x for feedback.

It appears that it cannot handle cyclic references. Probably worth a mention at least.

Reply:
There's nothing standard AFAIK, but we do have `objectEquals` in `src/common/util/util.ts` already. (Maybe not the easiest-to-find name.)

It doesn't handle cycles and most implementations of deep equality out there don't. It's also slightly simpler at least in function comparisons and probably some other things, so if you need them you can improve that implementation.

Reply:
I no longer need this, so removed it. There is also an implementation in src/common/util/util.ts that I should have been using.